### DISCUSSION TITLE
build.gradle: Patch version of spring-boot-gradle-plugin is not updated

### DISCUSSION THREAD

**Discussion by verbathim** [renovatebot/renovate#12201] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

28.1.4

### Describe the bug

The patch version of `org.springframework.boot:spring-boot-gradle-plugin` is not updated, e.g. from 2.4.10 to the latest 2.4.11.
I've disabled major and minor updates for Spring Boot, as they regularly break our application and need special manual treatment. I've explicitly allowed patch updates, but that does not lead to a patch update.
If I enable major/minor updates for Spring, a corresponding MR is created.

I've set the `LOG_LEVEL` to `debug` before running renovate. As a non-major update, only Spring Boot 2.5.5 is suggested, and no 2.4.11. I do not want to update to 2.5.5, but to 2.4.11 which is still a supported version.

https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-gradle-plugin/ seems to look fine w.r.t. v2.4.11.

I've included a minimal non-working example:

**config.js**

> module.exports = {
>   "baseBranches": ["develop"],
>   "platform": "gitlab",
>   "endpoint": "https://git.foo.bar/api/v4",
>   "gradle": { "enabled": true },
>   "gradle-wrapper": { "enabled": false },
>   "packageRules": [
>     {
>       "matchPackagePrefixes": ["org.springframework"],
>       "matchUpdateTypes": ["major", "minor"],
>       "enabled": false
>     },
>     {
>       "matchPackagePrefixes": ["org.springframework"],
>       "matchUpdateTypes": ["patch"],
>       "enabled": true
>     }
>   ]
> }

**build.gradle**

> buildscript {
>     ext {
>         springBootVersion = '2.4.10'
>     }
>     dependencies {
>         classpath "org.springframework.boot:spring-boot-gradle-plugin:${springBootVersion}"
>         classpath 'se.transmode.gradle:gradle-docker:1.2'
>     }
> }
> 
> project.version = versioningStrategy(0, 3)
> project.group = 'foo'
> project.ext.tenant = 'bar'
> 
> 
> apply plugin: 'java'
> apply plugin: 'idea'
> apply plugin: 'org.springframework.boot'
> 
> [...]

**Relevant output of renovate with LOG_LEVEL=debug**

> "gradle": [
>   {
>     "packageFile": "build.gradle",
>     "datasource": "maven",
>     "deps": [
>       {
>         "depName": "org.springframework.boot:spring-boot-gradle-plugin",
>         "currentValue": "2.4.10",
>         "managerData": {
>           "fileReplacePosition": 53,
>           "packageFile": "build.gradle"
>         },
>         "groupName": "springBootVersion",
>         "fileReplacePosition": 53,
>         "registryUrls": ["https://repo.maven.apache.org/maven2"],
>         "depIndex": 0,
>         "updates": [
>           {
>             "bucket": "non-major",
>             "newVersion": "2.5.5",
>             "newValue": "2.5.5",
>             "releaseTimestamp": "2021-09-23T08:05:15.000Z",
>             "newMajor": 2,
>             "newMinor": 5,
>             "updateType": "minor",
>             "branchName": "renovate/springbootversion"
>           }
>         ],
>         "warnings": [],
>         "versioning": "gradle",
>         "sourceUrl": "https://github.com/spring-projects/spring-boot",
>         "homepage": "https://spring.io/projects/spring-boot",
>         "currentVersion": "2.4.10",
>         "isSingleVersion": true,
>         "fixedVersion": "2.4.10"
>       },
>       [...]

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment, marked as answer]
https://docs.renovatebot.com/configuration-options/#separateminorpatch

**Comment by verbathim** [comment]
Wow, thanks, but I think this should be documented way better, since not every project maintainer respects semVer.
Why is this not a default?

**Reply by rarkins** [reply]
Because most people don't want that behavior.

> Wow, thanks, but I think this should be documented way better, since not every project maintainer respects semVer. 

![image](https://user-images.githubusercontent.com/6311784/137752808-10b5e2aa-6ebe-4d3f-bd4f-7fb4d2a0e2bc.png)

I think it's documented fairly clearly, but you are welcome to submit a PR to this repo improving it. Everything is under docs/

> Why is this not a default?

Most users don't want to separate minor from patch.

**Reply by verbathim** [reply]
I think at least some warning or hint should be produced by renovate. If I have to disable major and minor updates for a project that obviously does not respect semVer (like Spring Boot), I accidentally and very easily "disable" also patch updates without noticing, which can be a huge security issue. I don't like that.

**Reply by verbathim** [reply]
Any reaction to this?

**Reply by rarkins** [reply]
https://github.com/renovatebot/renovate/pull/12211/files#diff-d2c3390ae22ad0ab15473e88bf0b676e44d2e1394cbbba07024a6f8a5dbb444cR34-R43

**Reply by verbathim** [reply]
Thanks for the post.
I don't think it's mainly a problem of your documentation, but w.r.t. to the behaviour of your software / missing WARN logs.

**Reply by rarkins** [reply]
Renovate's flexibility is by nature too complex to catch and warn about every user misconfiguration. Expecting so is as unrealistic as expecting Apache or Terraform to themselves catch every possible misconfiguration. You're welcome to your opinion but unless you plan to start contributing to the project too then please stop causing us to spend time responding to your complaints multiple times.
